modules/frontend/search_streaming.go

```diff
@@ -0,0 +1,195 @@
+package frontend
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/tempo/modules/overrides"
+	"github.com/grafana/tempo/pkg/api"
+	"github.com/grafana/tempo/pkg/tempopb"
+	"github.com/grafana/tempo/tempodb"
+	"github.com/pkg/errors"
+)
+
+// diffSearchProgress only returns new and updated traces when result() is called
+// it uses a wrapped searchProgress to do all of the normal tracking as well as a map
+// to track if a trace was updated or not
+type diffSearchProgress struct {
+	progress shardedSearchProgress
+
+	seenTraces map[string]struct{}
+	mtx        sync.Mutex
+}
+
+func newDiffSearchProgress(ctx context.Context, limit, totalJobs, totalBlocks, totalBlockBytes int) shardedSearchProgress {
+	return &diffSearchProgress{
+		seenTraces: map[string]struct{}{},
+		progress:   newSearchProgress(ctx, limit, totalJobs, totalBlocks, totalBlockBytes),
+	}
+}
+
+func (p *diffSearchProgress) setStatus(statusCode int, statusMsg string) {
+	p.progress.setStatus(statusCode, statusMsg)
+}
+
+func (p *diffSearchProgress) setError(err error) {
+	p.progress.setError(err)
+}
+
+func (p *diffSearchProgress) addResponse(res *tempopb.SearchResponse) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	// record modified traces
+	for _, trace := range res.Traces {
+		p.seenTraces[trace.TraceID] = struct{}{}
+	}
+	p.progress.addResponse(res)
+}
+
+// shouldQuit locks and checks if we should quit from current execution or not
+func (p *diffSearchProgress) shouldQuit() bool {
+	return p.progress.shouldQuit()
+}
+
+func (p *diffSearchProgress) result() *shardedSearchResults {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	res := p.progress.result()
+
+	// filter result down to only traces in seenTraces
+	if res.response != nil {
+		keepTraces := make([]*tempopb.TraceSearchMetadata, 0, len(res.response.Traces))
+		for _, trace := range res.response.Traces {
+			_, ok := p.seenTraces[trace.TraceID]
+			if ok {
+				keepTraces = append(keepTraces, trace)
+			}
+		}
+		res.response.Traces = keepTraces
+	}
+	// clear seen traces
+	p.seenTraces = map[string]struct{}{}
+
+	return res
+}
+
+// finalResult gives the user the ability to pull all results w/o filtering
+// to ensure that all results are sent to the caller
+func (p *diffSearchProgress) finalResult() *shardedSearchResults {
+	return p.progress.result()
```

Comment:
Thinking through how a downstream system will consume the results (Grafana) - we might need to tweak something to resolve some ambiguity. Ex: if we stream results X, then Y, they should be added to the UI immediately. But then the final results X,Y,Z come in, it either has to replace the whole results or do its own diff to detect Z. Also search results for XY might change if new partials are found, and the new values would be streamed again.    I am thinking we could add a flag to denote the "final" results, or maybe there is some other mechanic that I'm missing?

Reply:
The contract that I'm seeing with the frontend is this:

If we return a traceid in the resultset then Tempo guarantees it is the complete set of results for that trace. The frontend can then replace any results it has that match that trace id. This will come into play more with grouping.

Reply:
Just wondering, do we need a finished state?  If the streaming were bidirectional the caller could decide what what was long enough to wait, or cancel the reuqest for whatever reason.

Reply:
> do we need a finished state?

Yeah now I'm thinking we don't need to send everything at the end.  The final diffs if any would be sufficient.

Reply:
After talking offline, we decided there isn't a straightforward path here and it's ok as-is.  Some things that need to be considered, not exhaustive:  bounded wait group .Add/.Done block because they don't take contexts, the http roundtripping can have other non-context related errors like json marshalling.